File: tech-js-orientadoObjetos/connect4/views/TurnView.js
import { PlayerVisitor } from '../models/Player.js';
import { HumanPlayerView, RandomPlayerView } from './PlayerView.js';
import { InIntervalDialog } from '../utils/views/Dialog.js';
import { Message } from './Message.js';



class TurnView extends PlayerVisitor {
    #turn;
    #activePlayerView;

    constructor(turn) {
        super();
        this.#turn = turn;
    }
    setNumberOfHumanPlayers() { 
        let inIntervalDialog = new InIntervalDialog(0, this.#turn.getNumberPlayers());
        inIntervalDialog.read(Message.NUM_PLAYERS.toString());
        this.#turn.reset(inIntervalDialog.getAnswer());
    }

    play() {
        this.#turn.getActivePlayer().accept(this);
        this.#turn.play(this.#activePlayerView.getColumn());
    }

    writeResult() {
        if ((this.#turn.getBoard()).isWinner()) {
            this.#activePlayerView.writeWinner();
        } else {
            Message.PLAYERS_TIED.writeln();
        }
    }
    visitHumanPlayer(humanPlayer) {
        this.#activePlayerView = new HumanPlayerView(humanPlayer);
    }
    visitRandomPlayer(randomPlayer) {
        this.#activePlayerView = new RandomPlayerView(randomPlayer);
    }
}

export { TurnView }

File: tech-java-orientadoObjetos/src/main/es/pbover/connect4Old/views/TurnView.java
package main.es.pbover.connect4Old.views;

import main.es.pbover.connect4Old.models.HumanPlayer;
import main.es.pbover.connect4Old.models.MinMaxPlayer;
import main.es.pbover.connect4Old.models.PlayerVisitor;
import main.es.pbover.connect4Old.models.RandomPlayer;
import main.es.pbover.connect4Old.models.Turn;
import main.es.pbover.utils.ClosedIntervalDialog;

public class TurnView implements PlayerVisitor {
    private Turn turn;
    private PlayerView activePlayerView;

    public TurnView(Turn turn) {
        super();
        this.turn = turn;
    }
    public void getNumberOfHumanPlayers() { 
        ClosedIntervalDialog closedIntervalDialog = new ClosedIntervalDialog(0, this.turn.getNumberPlayers());
        closedIntervalDialog.read(Message.NUM_PLAYERS.toString());
        this.turn.reset(closedIntervalDialog.getAnswer());
    }
    public void getMachineTypePlayer() { 
        ClosedIntervalDialog closedIntervalDialog = new ClosedIntervalDialog(1, 2);
        closedIntervalDialog.read(Message.TYPE_MACHINE.toString());
        this.turn.setTypeMachine(closedIntervalDialog.getAnswer());
    }

    public void play() {
        this.turn.getActivePlayer().accept(this);
        this.turn.play(this.activePlayerView.getColumn());
    }

    public void writeResult() {
        if ((this.turn.getBoard()).isWinner()) {
            this.activePlayerView.writeWinner();
        } else {
            Message.PLAYERS_TIED.writeln();
        }
    }
    public void visit(HumanPlayer humanPlayer) {
        this.activePlayerView = new HumanPlayerView(humanPlayer);
    }
    public void visit(RandomPlayer randomPlayer) {
        this.activePlayerView = new RandomPlayerView(randomPlayer);
    }
    public void visit(MinMaxPlayer minMaxPlayer) {
        this.activePlayerView = new MinMaxPlayerView(minMaxPlayer);
    }
}
